### PR TITLE
Decode a length-delimited framing protocol.

### PR DESCRIPTION
**Pull request by blossomwoo**
Length-delimited frame decoder. (#474)

### Motivation:

It would be useful to have a general ChannelInboundHandler that can decode a length-delimited framing protocol. This is a protocol that has messages prefixed with an encoding of their length.

Such a protocol would look like this:

`<fixed width integer n in network byte order><n bytes>
`

The simplest version of this would be one that uses a fixed-width integer.

For now we won't bother with varint support.

### Modifications:

Added a length-delimited frame decoder.

### Result:

People can now decode length-delimited framing protocols.

### CODE REVIEW

File: Sources/NIO/FixedLengthFrameDecoder.swift [hunk 1]
```diff
@@ -0,0 +1,47 @@
...
+// Licensed under Apache License v2.0
+//
+// See LICENSE.txt for license information
+// See CONTRIBUTORS.txt for the list of SwiftNIO project authors
+//
+// SPDX-License-Identifier: Apache-2.0
+//
+//===----------------------------------------------------------------------===//
+
+public final class FixedLengthFrameDecoder<T: FixedWidthInteger>: ByteToMessageDecoder {
```
Comment (normanmaurer) [inline Sources/NIO/FixedLengthFrameDecoder.swift]: I think we should rename this to `LengthFieldBasedFrameDecoder` as its not really fixes but depends on the length that is used to prefix the actual data.

File: Sources/NIO/FixedLengthFrameDecoder.swift [hunk 2]
```diff
@@ -0,0 +1,47 @@
...
+                return .needMoreData
+            }
+            state = .data(Int(integer))
+            return .continue
+        case .data(let dataLength):
+            guard let bytes = buffer.readSlice(length: dataLength) else {
+                return .needMoreData
+            }
+            ctx.fireChannelRead(self.wrapInboundOut(bytes))
+            state = .len
```
Comment (normanmaurer) [inline Sources/NIO/FixedLengthFrameDecoder.swift]: I think we should update the state before we call `ctx.fireChannelRead` to better protected of re-entrance.

Comment (normanmaurer) [inline Sources/NIO/FixedLengthFrameDecoder.swift]: also please use `self.state` (that's what we usually do).

File: Sources/NIO/FixedLengthFrameDecoder.swift [hunk 3]
```diff
@@ -0,0 +1,47 @@
...
+    
+    public init() { }
+    
+    public func decode(ctx: ChannelHandlerContext, buffer: inout ByteBuffer) -> DecodingState {
+        switch state {
+        case .len:
+            guard let integer = buffer.readInteger(as: T.self) else {
+                return .needMoreData
+            }
+            state = .data(Int(integer))
```
Comment (normanmaurer) [inline Sources/NIO/FixedLengthFrameDecoder.swift]: please use self.state

File: Sources/NIO/FixedLengthFrameDecoder.swift [hunk 4]
```diff
@@ -0,0 +1,47 @@
...
+    
+    enum State {
+        case len
+        case data(Int)
+    }
+    
+    public init() { }
+    
+    public func decode(ctx: ChannelHandlerContext, buffer: inout ByteBuffer) -> DecodingState {
+        switch state {
```
Comment (normanmaurer) [inline Sources/NIO/FixedLengthFrameDecoder.swift]: self.state

File: Tests/NIOTests/CodecTest.swift [hunk 5]
```diff
@@ -377,4 +377,76 @@ public class MessageToByteEncoderTest: XCTestCase {
         XCTAssertFalse(try channel.finish())
 
     }
+    
+    struct fixedLengthFrameDecoderHelper<T: FixedWidthInteger> {
```
Comment (normanmaurer) [inline Tests/NIOTests/CodecTest.swift]: I think I would move the tests to an extra file as the codec itself is also in an extra file.

File: Tests/NIOTests/CodecTest.swift [hunk 6]
```diff
@@ -377,4 +377,76 @@ public class MessageToByteEncoderTest: XCTestCase {
...
+        static fileprivate func fixedLengthFrameHelper(_ channel: EmbeddedChannel) throws {
+            _ = try channel.pipeline.add(handler: FixedLengthFrameDecoder<T>()).wait()
```
Comment (normanmaurer) [inline Tests/NIOTests/CodecTest.swift]: Use `XCTAssertNotThrows(...)`

File: Tests/NIOTests/CodecTest.swift [hunk 7]
```diff
@@ -377,4 +377,76 @@ public class MessageToByteEncoderTest: XCTestCase {
...
+            
+            for testString in testStrings {
+                channel.pipeline.fireChannelRead(NIOAny(buffer))
+                
+                guard var result: ByteBuffer = channel.readInbound() else {
+                    XCTFail("Failed to read buffer for \(T.self)")
+                    return
+                }
+                let testResult = result.readString(length: (result.readableBytes))
+                XCTAssert(testResult == testString, "String doesn't match.")
```
Comment (normanmaurer) [inline Tests/NIOTests/CodecTest.swift]: Use `XCTAssertEqual(...)`

File: Tests/NIOTests/CodecTest.swift [hunk 8]
```diff
@@ -377,4 +377,76 @@ public class MessageToByteEncoderTest: XCTestCase {
...
+            
+            let testStrings = ["try", "swift", "san", "jose"]
+            var buffer = channel.allocator.buffer(capacity:1024)
+            for testString in testStrings {
+                buffer.write(integer: T(testString.count))
+                buffer.write(string: testString)
+            }
+            
+            for testString in testStrings {
+                channel.pipeline.fireChannelRead(NIOAny(buffer))
```
Comment (normanmaurer) [inline Tests/NIOTests/CodecTest.swift]: use `channel.writeInbound(...)` and assert the return value to replace `fireChannelRead` calls.

File: Tests/NIOTests/CodecTest.swift [hunk 9]
```diff
@@ -377,4 +377,76 @@ public class MessageToByteEncoderTest: XCTestCase {
...
+                
+                guard var result: ByteBuffer = channel.readInbound() else {
+                    XCTFail("Failed to read buffer for \(T.self)")
+                    return
+                }
+                let testResult = result.readString(length: (result.readableBytes))
+                XCTAssert(testResult == testString, "String doesn't match.")
+            }
+        }
```
Comment (normanmaurer) [inline Tests/NIOTests/CodecTest.swift]: call `channel.finish()` and assert return value.

File: Sources/NIO/FixedLengthFrameDecoder.swift (hunk 1 above)
Comment (normanmaurer) [inline Sources/NIO/FixedLengthFrameDecoder.swift]: Can you please add some docs, something like this:

https://github.com/netty/netty/blob/4.1/codec/src/main/java/io/netty/handler/codec/LengthFieldBasedFrameDecoder.java#L184

Just a bit simplified as this implementation always assume the length is on offset 0.

File: Sources/NIO/LengthFieldBasedFrameDecoder.swift [hunk 10]
```diff
@@ -0,0 +1,61 @@
...
+/// ````
+public final class LengthFieldBasedFrameDecoder<T: FixedWidthInteger>: ByteToMessageDecoder {
+    public typealias InboundIn = ByteBuffer
+    public typealias InboundOut = ByteBuffer
+    
+    public var cumulationBuffer: ByteBuffer?
+    
+    private var state: State = .len
+    
+    enum State {
```
Comment (weissi) [inline Sources/NIO/LengthFieldBasedFrameDecoder.swift]: can we make this `private enum State`?

File: Sources/NIO/LengthFieldBasedFrameDecoder.swift [hunk 11]
```diff
@@ -0,0 +1,61 @@
...
+        case len
```
Comment (weissi) [inline Sources/NIO/LengthFieldBasedFrameDecoder.swift]: I'd argue for making those cases more descriptive. Like `case waitingForLength` and `case waitingForPayload(Int)`. @normanmaurer / @Lukasa  what do you think?

File: Sources/NIO/LengthFieldBasedFrameDecoder.swift [hunk 12]
```diff
@@ -0,0 +1,61 @@
...
+        case data(Int)
+    }
+    
+    public init() { }
```
Comment (weissi) [inline Sources/NIO/LengthFieldBasedFrameDecoder.swift]: this definitely needs an upper bound parameter. Otherwise an attacker could trivially kill this server by sending a massive length.

File: Sources/NIO/LengthFieldBasedFrameDecoder.swift [hunk 13]
```diff
@@ -0,0 +1,61 @@
...
+    
+    public func decode(ctx: ChannelHandlerContext, buffer: inout ByteBuffer) -> DecodingState {
+        switch self.state {
+        case .len:
+            guard let integer = buffer.readInteger(as: T.self) else {
+                return .needMoreData
+            }
+            self.state = .data(Int(integer))
```
Comment (weissi) [inline Sources/NIO/LengthFieldBasedFrameDecoder.swift]: this needs to verify a few things:

- `> 0`
- `<= self.upperBound`

File: Sources/NIO/LengthFieldBasedFrameDecoder.swift [hunk 14]
```diff
@@ -0,0 +1,61 @@
...
+        case .data(let dataLength):
+            guard let bytes = buffer.readSlice(length: dataLength) else {
+                return .needMoreData
+            }
+            self.state = .len
+            ctx.fireChannelRead(self.wrapInboundOut(bytes))
+            return .continue
+        }
+    }
+}
```
Comment (weissi) [inline Sources/NIO/LengthFieldBasedFrameDecoder.swift]: we should do something when the handler is removed and `self.cumulationBuffer` is non-empty. @Lukasa / @normanmaurer I reckon we should just fire an error through the pipeline is the handler is removed and we're left with some partial message, no?

File: Tests/NIOTests/LengthFieldBasedFrameDecoderTest.swift [hunk 15]
```diff
@@ -0,0 +1,91 @@
...
+// See CONTRIBUTORS.txt for the list of SwiftNIO project authors
+//
+// SPDX-License-Identifier: Apache-2.0
+//
+//===----------------------------------------------------------------------===//
+
+//
+import XCTest
+
+struct fixedLengthFrameDecoderHelper<T: FixedWidthInteger> {
```
Comment (weissi) [inline Tests/NIOTests/LengthFieldBasedFrameDecoderTest.swift]: - the type name should be upper case
- this helper type should go to the end of the file. In Xcode it's really helpful is the test class itself is at the top as it contains the 'run' icon for the whole test suite right next to the class. If it's not at the top it's always kind of hard to find the test class as you need to scroll and look for it...

File: Tests/NIOTests/LengthFieldBasedFrameDecoderTest.swift [hunk 16]
```diff
@@ -0,0 +1,91 @@
...
+// Licensed under Apache License v2.0
+//
+// See LICENSE.txt for license information
+// See CONTRIBUTORS.txt for the list of SwiftNIO project authors
+//
+// SPDX-License-Identifier: Apache-2.0
+//
+//===----------------------------------------------------------------------===//
+
+//
```
Comment (weissi) [inline Tests/NIOTests/LengthFieldBasedFrameDecoderTest.swift]: please remove the `//` line

File: Tests/NIOTests/LengthFieldBasedFrameDecoderTest.swift [hunk 17]
```diff
@@ -0,0 +1,91 @@
...
+            let testResult = result.readString(length: (result.readableBytes))
+            XCTAssertEqual(testResult, testString)
+        }
+        XCTAssertFalse(try channel.finish())
+    }
+}
+
+public class LengthFieldBasedFrameDecoderTest: XCTestCase {
+    func testFixedLengthFrameDecoderInt8() throws {
+        let channel = EmbeddedChannel()
```
Comment (weissi) [inline Tests/NIOTests/LengthFieldBasedFrameDecoderTest.swift]: the `EmbeddedChannel` could just be created in `fixedLengthFrameHelper` helper function, no?

File: Tests/NIOTests/LengthFieldBasedFrameDecoderTest.swift [hunk 18]
```diff
@@ -0,0 +1,91 @@
...
+                return
+            }
+            let testResult = result.readString(length: (result.readableBytes))
+            XCTAssertEqual(testResult, testString)
+        }
+        XCTAssertFalse(try channel.finish())
+    }
+}
+
+public class LengthFieldBasedFrameDecoderTest: XCTestCase {
```
Comment (weissi) [inline Tests/NIOTests/LengthFieldBasedFrameDecoderTest.swift]: we need tests for the edge cases, for example:

- negative lengths
- absurdly large lengths
- length and then before receiving the appropriate number of bytes we get an EOF
- EOF whilst waiting for length

Comment (blossomwoo) [reply Tests/NIOTests/LengthFieldBasedFrameDecoderTest.swift]: @weissi I'm having trouble with the last 2 tests. Could you take a look at `testEOFBeforeReceivingAppropriateNumberOfBytes` & `testEOFWhileWaitingForLength` and provide some pointers? Thanks!

File: Sources/NIO/LengthFieldBasedFrameDecoder.swift [hunk 19]
```diff
@@ -0,0 +1,85 @@
...
+//
+// SPDX-License-Identifier: Apache-2.0
+//
+//===----------------------------------------------------------------------===//
+
+/// A `ChannelInboundHandler` used to decode length-delimited frames.
+///
+/// This is a protocol that has messages prefixed with an encoding of their length.
+/// Such a protocol would look like this:
+/// ````
```
Comment (weissi) [inline Sources/NIO/LengthFieldBasedFrameDecoder.swift]: - isn't this one backtick too many?
- super minor: could we have an empty line (just `///`) between the code blocks?

File: Sources/NIO/LengthFieldBasedFrameDecoder.swift [hunk 20]
```diff
@@ -0,0 +1,85 @@
...
+    private var state: State = .waitingForLength
+    
+    /// Errors thrown by the NIO LengthFieldBasedFrameDecoder module.
+    public enum NIOLengthFieldBasedFrameDecoderError: Error {
+        /// The frame being sent is negative length or larger than the configured maximum
+        /// acceptable frame size
+        case invalidFrameLength
+        
+        /// when the handler is removed and buffer is non-empty
+        case bytesLeftOver
```
Comment (weissi) [inline Sources/NIO/LengthFieldBasedFrameDecoder.swift]: this should have the bytes that were left over as associated data

File: Sources/NIO/LengthFieldBasedFrameDecoder.swift [hunk 21]
```diff
@@ -0,0 +1,85 @@
...
+        self.upperBound = upperBound
+    }
+    
+    public func decode(ctx: ChannelHandlerContext, buffer: inout ByteBuffer) throws -> DecodingState {
+        switch self.state {
+        case .waitingForLength:
+            guard let integer = buffer.readInteger(as: T.self) else {
+                return .needMoreData
+            }
+            guard integer > 0 && integer <= self.upperBound else {
```
Comment (weissi) [inline Sources/NIO/LengthFieldBasedFrameDecoder.swift]: this should be `>= 0`, no? I think `0` is fine no?

File: Tests/NIOTests/LengthFieldBasedFrameDecoderTest.swift (hunk 18 above)
Comment (weissi) [reply Tests/NIOTests/LengthFieldBasedFrameDecoderTest.swift]: hi @blossomwoo sure, that's not obvious at all actually. Let me run you through what I think we need to do to get this in asap. We're really close!

If anything doesn't make sense, please just at-mention me and ask for help. And sorry this is quite a bit of stuff actually.

So first we should make some very minor changes to the actual code:

1) we should create a new `private func handleLeftOverData()` which contains the logic to fire the error if there's left-over stuff
2) we should call `self.handleLeftOverData()` from both `func handlerRemoved()` and `func channelInactive()`

The reason we should do it in both `channelInactive` (called when the channel gets inactive, in TCP that means the socket is closed) and `handlerRemoved` (called when the handler gets removed from the pipeline) is: In probably most cases, you'll see `channelInactive` followed by `handlerRemoved` because we remove all handlers because we tear the pipeline down. However, a handler might also be removed in normal operation in which case we would also like to fire the error. The easiest way of achieving this is to call the left-over logic in both. Note: we need to set the left-over bytes to nothing after firing the error.

regarding the tests, what probably works best is to create a new channel handler called something like

    private final class LeftOverCheckHandler: ChannelInboundHandler {
        typealias InboundIn = ByteBuffer
        typealias InboundOut = ByteBuffer

        enum LeftOverCheckHandlerError: Error {
            case wrongErrorReceived(Error)
        }

        private var haveSeenLeftOvers = false
        private let leftOversReceivedPromise: EventLoopPromise<ByteBuffer>

        init(leftOversReceivedPromise: EventLoopPromise<ByteBuffer >) {
            self.leftOversReceivedPromise = leftOversReceivedPromise
        }

        func errorCaught(ctx: ChannelHandlerContext, error: Error) {
            XCTAssertFalse(self.haveSeenLeftOvers)
            self.haveSeenLeftOvers = true

            switch error {
                case let error as NIOLengthFieldBasedFrameDecoderError:
                    switch error {
                        case .bytesLeftOver(let bytes):
                            self.leftOversReceivedPromise.succeed(value: bytes)
                        case .invalidFrameLength:
                            self.leftOversReceivedPromise.fail(error: LeftOverCheckHandlerError.wrongErrorReceived(error))
                    }
                  default:
                    self.leftOversReceivedPromise.fail(error: LeftOverCheckHandlerError.wrongErrorReceived(error))
                }
         }
     }

and you can just add this channel handler to the channel (using the following code)

```
let leftOversReceivedPromise: EventLoopPromise<ByteBuffer> = embedded.eventLoop.newPromise()
XCTAssertNoThrow(try embedded.pipeline.add(handler: LeftOverCheckHandler(leftOversReceivedPromise: leftOversReceivedPromise)).wait())
```

and now you can do a bunch of test cases:

1) just fire `channelInactive` through the pipeline manually: `embedded.pipeline.fireChannelInactive()` then verify that the left over bytes got received correctly by the `LeftOverCheckHandler`:

    XCTAssertEqual(whateverLeftOverBytesYouExpectAsByteBuffer,
                                 leftOversReceivedPromise.futureResult.wait()!)

2) call `embedded.close().wait()` and then spin the event loop (`embedded.eventloop.run()`) and then verify the same thing as in (1)
3) left the embedded channel up and running but remove the handler (`embedded.pipeline.remove(handler: theHandlerInstanceOfYourFixedLengthDecoderHandler)`) and then verify the same thing as in (1) and (2)

(disclaimer: I tested/compiled none of the code in here so there might be some minor mistakes, bugs, typos, etc)

File: Sources/NIO/Linux.swift [hunk 22]
```diff
@@ -103,9 +103,9 @@ internal enum Linux {
     static let SOCK_NONBLOCK = Int32(bitPattern: Glibc.SOCK_NONBLOCK.rawValue)
```
Comment (weissi) [inline Sources/NIO/Linux.swift]: please also remove all the changes in this file, I think Xcode's 'rename' function got properly confused here...

File: Sources/NIO/Selector.swift [hunk 23]
```diff
@@ -297,13 +297,13 @@ final class Selector<R: Registration> {
```
Comment (weissi) [inline Sources/NIO/Selector.swift]: please also remove all the changes in this file, I think Xcode's 'rename' function got properly confused here...

**Comment by Lukasa** [comment]
I think this is now covered by apple/swift-nio-extras#17.
